src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs

```diff
@@ -19,86 +19,149 @@ namespace Microsoft.CodeAnalysis.PdbSourceDocument
-                        if (fileChecksum.SequenceEqual(sourceDocument.Checksum))
+                        stream.Position = 0;
+                        using (var file = File.OpenWrite(filePath))
                         {
-                            var textAndVersion = TextAndVersion.Create(sourceText, VersionStamp.Default, sourceDocument.FilePath);
-                            var textLoader = TextLoader.From(textAndVersion);
-                            return Task.FromResult<TextLoader?>(textLoader);
+                            stream.CopyTo(file);
                         }
+
+                        new FileInfo(filePath).IsReadOnly = true;
                     }
                     catch (IOException)
                     {
                         // TODO: Log message to inform the user what went wrong: https://github.com/dotnet/roslyn/issues/57352
+                        return null;
                     }
                 }
-            }
 
-            // TODO: Call the debugger to download the file
-            // Maybe they'll download to a temp file, in which case this method could return a string
-            // or maybe they'll return a stream, in which case we could create a new StreamTextLoader
+                return LoadSourceFile(filePath, sourceDocument, encoding, ignoreChecksum: false);
+            }
 
-            return Task.FromResult<TextLoader?>(null);
+            return null;
         }
 
-        private static Stream? TryGetEmbeddedSourceStream(SourceDocument sourceDocument)
+        private async Task<SourceFileInfo?> TryGetSourceLinkFileAsync(SourceDocument sourceDocument, Encoding encoding, IPdbSourceDocumentLogger? logger, CancellationToken cancellationToken)
         {
-            if (sourceDocument.EmbeddedTextBytes is null)
+            if (_sourceLinkService is null || sourceDocument.SourceLinkUrl is null)
                 return null;
 
-            var embeddedTextBytes = sourceDocument.EmbeddedTextBytes;
-            var uncompressedSize = BitConverter.ToInt32(embeddedTextBytes, 0);
-            var stream = new MemoryStream(embeddedTextBytes, sizeof(int), embeddedTextBytes.Length - sizeof(int));
+            // This should ideally be the repo-relative path to the file, and come from SourceLink: https://github.com/dotnet/sourcelink/pull/699
+            var relativePath = Path.GetFileName(sourceDocument.FilePath);
 
-            if (uncompressedSize != 0)
-            {
-                var decompressed = new MemoryStream(uncompressedSize);
+            var delay = Task.Delay(SourceLinkTimeout, cancellationToken);
+            var sourceFileTask = _sourceLinkService.GetSourceFilePathAsync(sourceDocument.SourceLinkUrl, relativePath, logger, cancellationToken);
 
-                using (var deflater = new DeflateStream(stream, CompressionMode.Decompress))
+            var winner = await Task.WhenAny(sourceFileTask, delay).ConfigureAwait(false);
```

Comment:
Not a huge deal, but I could see this being useful in [TaskExtensions](https://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/TaskExtensions.cs,599b9667820e1f18)

Reply:
Something like, `Task.RunWithTimeout<T>(Task<T> taskToRun, int timeoutInMs, CancellationToken cancellationToken)`

Reply:
Will leave this for a follow up 👍